File: Java/LinkedList/01_Singly_Linear_LinkedList_Java/SinglyLL_003.java
/*
===============================================================================
Program Name    : Singly Linear Linked List - Accessing Node Data
Program Number  : 003
Description     : This program demonstrates how to access the data
                  stored in individual nodes of a Singly Linear
                  Linked List using the head reference and chained
                  next references.

                  Three node objects are dynamically created,
                  linked together, and their data members are
                  accessed sequentially without using loops.

Date            : July 2026
Language        : Java
Author          : Ishwari Vijaykumar Surve

===============================================================================
Concepts Used:

1. Classes and Objects
2. Self-Referential Class
3. Object References
4. Dynamic Object Creation using new
5. Head Reference
6. Manual Node Linking
7. Sequential Node Access

Key Learning Points:

- The head reference always points to the first node.
- Every node is accessed through its previous node.
- The next reference creates the connection between nodes.
- Chained references can be used to reach any node in the list.
- This program introduces the concept of linked list traversal without loops.

===============================================================================
*/

class node
{
    public int data;
    public node next;
}

class SinglyLL003
{
    public static void main(String A[])
    {
        node head = null;
        node obj1 = null;
        node obj2 = null;
        node obj3 = null;

        obj1 = new node();
        obj2 = new node();
        obj3 = new node();

        obj1.data = 11;
        obj2.data = 21;
        obj3.data = 51;

        obj1.next = obj2;
        obj2.next = obj3;
        obj3.next = null;

        head = obj1;

        System.out.println(head.data);
        System.out.println(head.next.data);
        System.out.println(head.next.next.data);
    }
}

/*
===============================================================================
Expected Output
===============================================================================

11
21
51

===============================================================================
Memory Diagram
===============================================================================

head
 |
 v
+-----------+     +-----------+     +-------------+
| data : 11 | --> | data : 21 | --> | data : 51   |
| next : ---|     | next : ---|     | next : null |
+-----------+     +-----------+     +-------------+

===============================================================================
Step-by-Step Execution
===============================================================================

Step 1:
Four reference variables are declared.

    node head = null;
    node obj1 = null;
    node obj2 = null;
    node obj3 = null;

-------------------------------------------------------------------------------

Step 2:
Memory is allocated for three node objects.

    obj1 = new node();
    obj2 = new node();
    obj3 = new node();

-------------------------------------------------------------------------------

Step 3:
Data values are assigned.

    obj1.data = 11;
    obj2.data = 21;
    obj3.data = 51;

-------------------------------------------------------------------------------

Step 4:
The nodes are linked together.

    obj1.next = obj2;
    obj2.next = obj3;
    obj3.next = null;

-------------------------------------------------------------------------------

Step 5:
The head reference stores the address of the first node.

    head = obj1;

-------------------------------------------------------------------------------

Step 6:
The node data is accessed through chained references.

First Node

    head.data

Output:

11

Second Node

    head.next.data

Output:

21

Third Node

    head.next.next.data

Output:

51

===============================================================================
Reference Flow
===============================================================================

head
 |
 v
obj1 ───► obj2 ───► obj3 ───► null

===============================================================================
Operations Performed
===============================================================================
1. Created a self-referential class.
2. Declared object references.
3. Allocated memory for three node objects.
4. Stored data in each node.
5. Linked all nodes using next references.
6. Assigned the first node to the head reference.
7. Accessed the first node using head.
8. Accessed the second node using head.next.
9. Accessed the third node using head.next.next.
10. Displayed the data stored in each node.

===============================================================================
Time Complexity
===============================================================================

Node Creation          : O(1)
Node Linking           : O(1)
Access First Node      : O(1)
Access Second Node     : O(1)
Access Third Node      : O(1)

Overall                : O(1)

Space Complexity       : O(1)

===============================================================================
Conclusion
===============================================================================

This program demonstrates how the head reference acts as the entry
point of a Singly Linear Linked List. By following the next references,
each node can be accessed sequentially. Understanding chained object
references is fundamental before implementing linked list traversal,
Display(), Count(), insertion, and deletion operations.

===============================================================================
*/

File: C++/LinkedList/01_Singly_Linear_LinkedList_CPP/SinglyLL_003.cpp
/*
===============================================================================
Program Name    : Introduction to Singly Linked List Class
Program Number  : 003 (C++ Version)
Description     : This program introduces the Singly Linked List class using
                  Object-Oriented Programming. It demonstrates constructor
                  invocation and initializes the head pointer (first) to NULL,
                  representing an empty linked list ready for future operations.
Date            : July 2026
Language        : C++
Author          : Ishwari Vijaykumar Surve
===============================================================================

Concepts Used:
1. Object-Oriented Programming (OOP)
2. Class and Object
3. Constructor
4. Self-Referential Structure
5. Pointer Initialization
6. Head Pointer (first)
7. this Pointer

Key Learning Points:
- Constructor executes automatically when an object is created.
- first acts as the head pointer of the linked list.
- first is initialized to NULL to represent an empty list.
- this pointer accesses the current object's data members.
- Forms the foundation for all linked list operations.

===============================================================================
*/

#include<iostream>
using namespace std;

#pragma pack(1)

struct node
{
    int data;
    struct node *next;
};

typedef struct node NODE;
typedef struct node *PNODE;
typedef struct node **PPNODE;

class SinglyLL
{
    public:

        PNODE first;

        /*
        =========================================================================
        Default Constructor

        - Automatically called whenever a SinglyLL object is created.
        - Initializes the head pointer (first) to NULL.
        - Indicates that the linked list is initially empty.
        =========================================================================
        */
        SinglyLL()
        {
            cout<<"Inside Constructor\n";
            this->first = NULL;
        }
};

int main()
{
    SinglyLL sobj;

    return 0;
}

/*
===============================================================================
Expected Output:
===============================================================================

Inside Constructor

===============================================================================
Memory Representation:
===============================================================================

             sobj
               │
               ▼
      +------------------+
      | first = NULL     |
      +------------------+

Linked List:

first
  │
  ▼
NULL

===============================================================================
Explanation:
===============================================================================

The program creates a Singly Linked List class containing a single data
member named first, which stores the address of the first node in the list.

When an object of the class is created, the constructor executes
automatically and initializes first to NULL. This indicates that the linked
list is empty because no nodes have been created yet.

This program focuses only on constructor-based initialization and serves as
the starting point for implementing linked list operations.

===============================================================================
Program Flow:
===============================================================================

Create SinglyLL Object
          │
          ▼
Constructor Invoked Automatically
          │
          ▼
Initialize first = NULL
          │
          ▼
Display "Inside Constructor"
          │
          ▼
Program Terminates

===============================================================================
Operations Performed:
===============================================================================

1. Defined a self-referential node structure.
2. Created the SinglyLL class.
3. Declared first as the head pointer.
4. Initialized first to NULL inside the constructor.
5. Created an object of the SinglyLL class.
6. Executed the constructor automatically.

===============================================================================
Time Complexity:
===============================================================================

Constructor      : O(1)
Space Complexity : O(1)

===============================================================================
Conclusion:
===============================================================================

Program 003 introduces the Singly Linked List class using Object-Oriented
Programming. The constructor initializes the head pointer to NULL, ensuring
that the list starts in an empty state. This serves as the foundation for
implementing insertion, deletion, traversal, and other linked list
operations in the upcoming programs.

===============================================================================
*/
